### PR TITLE
User - CRUD API 

### LINKED ISSUE #5: [Backend] User - CRUD API
# Objetive

Criar API de usuários, mantendo os cenários na descrição

## Description

A API de CRUD deve ter os seguintes requisitos, respeitando as especificações mencionadas em "Observações":

- [x] deve cadastrar um `user`;
- [x] deve editar um `user` passando um `id`;
- [x] deve realizar um _soft delete_ para um `user` passando um `id`;
- [x] deve ser possível buscar por:
  - [x] um `user` passando um `id`;
  - [x] vários `user`, contendo paginação:
    - podendo filtrar os itens por:
      - [x] um ou mais `user` passando **ou** `email` **ou** `username` por regex;
    - podendo ordenar os itens por:
      - [x] `username`, vem por padrão, ordem crescente;
      - [x] `createdAt`, vem por padrão, ordem decrescente;
    - informando os seguintes campos de paginação:
      - [x] campo `size`: o valor deve estar entre `5`, `10`, `15`, `20`, `50` ou `100`;
      - [x] campo `page`: o valor deve ser maior que `1`;
    - [x] retornando os seguintes campos:
      - campo `page`: página atual, padrão `1`;
      - campo `total`: total de registros, padrão `0`;
      - campo `size`: total de itens por `page`, padrão `10`;
      - campo `pageCount`: total de páginas de acordo com o `total` e `size`;
      - campo `hasPrevious`: informa se existe "página anterior", padrão `false`;
      - campo `hasNext`: informa se existe "próxima página", padrão `false`;
      - campo `data`: todos os resultados da pesquisa, padrão `[]`;
- [x] todas as rotas devem possuir OpenAPI (Swagger);
- [x] deve ter testes unitários em todas as rotas
- [x] deve ter um `coverage` de testes de `100%`;

### Observações

Os campos da entidade, deverão seguir às especificações do Wiki, seção _"User Fields"_ (link no final).

## Links

- [Modelos do Banco de Dados](https://github.com/never-stop-learning-br/clone-whatsapp-backend/wiki/Modelos-do-Banco-de-Dados)

### PR DESCRIPTION
**Pull request by gabrielcdiniz**
# Description 

This Pull Request, append the following resources:

* add dependencies:
  - add bcrypt
  - add class-validator
  - add class-transform
  - add @faker-js/faker
  - add @types/bcrypt
  - add eslint-plugin-import-helpers
  
* add interceptors:
  - "remove fields" interceptor

* add endpoints:
  - add user "createOne" method
  - add user "updateOneById" method

### CODE REVIEW

File: src/modules/users/dto/create-user.dto.ts [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+    maxLength: 256,
+  })
+  @MinLength(6)
+  @MaxLength(256)
+  @IsStrongPassword()
+  public password: string;
+
+  @IsOptional()
+  @IsDate()
+  public deletedAt: Date;
```
Comment (gabrielcdiniz) [inline src/modules/users/dto/create-user.dto.ts]: remove

File: src/modules/users/entities/user.entity.ts [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+  })
+  public email: string;
+
+  @Prop({
+    type: String,
+    required: [true, ERR_DB_REQUIRED],
+    minlength: [6, ERR_DB_MIN_6_CHARACTERS],
+    maxlength: [256, ERR_DB_MAX_256_CHARACTERS],
+  })
+  public password: string;
```
Comment (gabrielcdiniz) [inline src/modules/users/entities/user.entity.ts:69]: where is the "deletedAt" property ?

File: src/modules/users/users.controller.spec.ts [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+      providers: [...mockedProviders],
+    }).compile();
+
+    controller = module.get<UsersController>(UsersController);
+  });
+
+  describe('createOne', () => {
+    it('should create one user', async () => {
+      // ? ARRANGE
+
```
Comment (gabrielcdiniz) [inline src/modules/users/users.controller.spec.ts]: remove whitespace

File: src/modules/users/users.controller.spec.ts [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+    });
+  });
+
+  describe('softDeleteById', () => {
+    it('should make softDelete one user by id', async () => {
+      // ? ARRANGE
+      const id = faker.database.mongodbObjectId();
+      // ? ACT
+      await controller.softDeleteById(id);
+      // ? ASSERT
```
Comment (gabrielcdiniz) [inline src/modules/users/users.controller.spec.ts]: give some spaces between comments and code, following the pattern

File: src/modules/users/users.controller.ts [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+    @Body() updateUserDto: UpdateUserDTO,
+  ) {
+    return this.usersService.updateOneById(id, updateUserDto);
+  }
+
+  @Delete(':id')
+  @UseInterceptors(new RemoveFieldsInterceptor<User>(['password']))
+  public softDeleteById(@Param('id') id: string) {
+    return this.usersService.softDeleteById(id);
+  }
```
Comment (gabrielcdiniz) [inline src/modules/users/users.controller.ts:128]: need sawgger api documentation

File: src/modules/users/users.service.spec.ts [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+  });
+
+  describe('softDeleteByid', () => {
+    it('should make softDelete one user by id', async () => {
+      // ? ARRANGE
+      const id = faker.database.mongodbObjectId();
+      const dto = {
+        username: faker.person.firstName(),
+        email: faker.internet.email(),
+        deletedAt: null,
```
Comment (gabrielcdiniz) [inline src/modules/users/users.service.spec.ts]: remove "deletedAt" field

File: src/modules/users/users.service.spec.ts [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+      // ? ACT
+      const returnedValue = await service.softDeleteById(id);
+
+      // ? ASSERT
+      expect(userModelMock.findByIdAndUpdate).toHaveBeenCalledTimes(1);
+      expect(userModelMock.findByIdAndUpdate).toHaveBeenCalledWith(
+        id,
+        {
+          deletedAt: new Date(),
+        },
```
Comment (gabrielcdiniz) [inline src/modules/users/users.service.spec.ts]: create a function util, in the path `src/shared/tests/utils` and add to const the new date

File: src/modules/users/dto/find-one-user.dto.ts [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+  MinLength,
+} from 'class-validator';
+
+export class FindOneFilterDTO {
+  @ApiProperty({
+    type: String,
+    description: 'Unique "username" in the application',
+    minLength: 4,
+    maxLength: 32,
+  })
```
Comment (gabrielcdiniz) [inline src/modules/users/dto/find-one-user.dto.ts]: should be "optional"

File: src/modules/users/dto/find-one-user.dto.ts [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+  @IsAlphanumeric()
+  @IsString()
+  public username?: string;
+
+  @ApiProperty({
+    type: String,
+    description: 'Unique "email" in the application',
+    minLength: 8,
+    maxLength: 256,
+  })
```
Comment (gabrielcdiniz) [inline src/modules/users/dto/find-one-user.dto.ts]: should be "optional"

File: src/modules/users/dto/find-one-user.dto.ts [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+  @MinLength(8)
+  @MaxLength(256)
+  @IsEmail()
+  public email?: string;
```
Comment (gabrielcdiniz) [inline src/modules/users/dto/find-one-user.dto.ts]: should add "@IsOptional()" decorator

File: src/modules/users/dto/find-one-user.dto.ts [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+    type: String,
+    description: 'Unique "username" in the application',
+    minLength: 4,
+    maxLength: 32,
+  })
+  @MinLength(4)
+  @MaxLength(32)
+  @IsAlphanumeric()
+  @IsString()
+  public username?: string;
```
Comment (gabrielcdiniz) [inline src/modules/users/dto/find-one-user.dto.ts]: should add "@IsOptional()" decorator

Comment (gabrielcdiniz) [inline src/modules/users/dto/find-one-user.dto.ts]: should add `readonly` key

File: src/modules/users/dto/find-one-user.dto.ts (hunk 10 above)
Comment (gabrielcdiniz) [inline src/modules/users/dto/find-one-user.dto.ts]: should add `readonly` key

File: src/modules/users/users-sort-validation/users-sort-validation.pipe.spec.ts [hunk 12]
```diff
@@ -0,0 +1,7 @@
+import { UsersSortValidationPipe } from './users-sort-validation.pipe';
+
+describe('UsersSortValidationPipe', () => {
+  it('should be defined', () => {
+    expect(new UsersSortValidationPipe()).toBeDefined();
+  });
+});
```
Comment (gabrielcdiniz) [inline src/modules/users/users-sort-validation/users-sort-validation.pipe.spec.ts]: should test it

File: src/modules/users/users.service.spec.ts [hunk 13]
```diff
@@ -0,0 +1,426 @@
...
+        password: faker.internet.password(),
+      };
+
+      const size = faker.number.int({ min: 5 });
+      const page = faker.number.int({ min: 1 });
+
+      const sort: SortUserDTO = {};
+      const filter: FindUserDTO = {
+        username: user.username,
+        email: user.username,
```
Comment (gabrielcdiniz) [inline src/modules/users/users.service.spec.ts]: should be "user.email" instead

File: src/modules/users/users.service.spec.ts [hunk 14]
```diff
@@ -0,0 +1,426 @@
...
+
+      const query: FilterQuery<User> = {};
+      const options: QueryOptions<User> = {
+        limit: size,
+        skip: (page - 1) * size,
+        sort,
+      };
+
+      const data = [user, user, user];
+      const total = faker.number.int({ min: 0, max: 50 });
```
Comment (gabrielcdiniz) [inline src/modules/users/users.service.spec.ts]: should remove "max" property

File: src/modules/users/users-sort-validation/users-sort-validation.pipe.spec.ts (hunk 12 above)
Comment (DhyonKeyllon) [reply src/modules/users/users-sort-validation/users-sort-validation.pipe.spec.ts]: shoulde be defined?

**Review by DhyonKeyllon — APPROVED** [review]

**Pull request merged**